### PR TITLE
Throw a graceful error when type checking for `ctpop` fails

### LINKED ISSUE #2121: Kani crashes when an intrinsic call using ZST is invalid
I tried this code:

```rust
#![feature(core_intrinsics)]
use std::intrinsics::ctpop;

// These shouldn't compile.
#[kani::proof]
pub fn check_zst_ctpop() {
    let n = ctpop(());
    assert!(n == ());
}
```

and also this code:

```rust
#![feature(core_intrinsics)]
use std::intrinsics::sub_with_overflow;

#[kani::proof]
pub fn check_zst_sub_with_overflow() {
    let n = sub_with_overflow((), ());
    assert!(!n.1);
}
```

using the following command line invocation:

```
kani ice.rs
```

with Kani version:

I expected to see this happen: Kani should fail with a compilation error similar to running `rustc ice.rs`

<!--
If Kani crashed, please include a backtrace in the code block by setting
`RUST_BACKTRACE=1` in your environment. E.g. `RUST_BACKTRACE=1 kani ...`
-->
Instead, this happened: Kani crashes.

```
thread '<unnamed>' panicked at 'BinaryOperation Expression does not typecheck OverflowResultMinus Expr { value: Symbol { identifier: "arithmetic_zst_fixme::check_zst_sub_with_overflow::1::var_2" }, typ: StructTag("tag-Unit"), loc
ation: None } Expr { value: Symbol { identifier: "arithmetic_zst_fixme::check_zst_sub_with_overflow::1::var_3" }, typ: StructTag("tag-Unit"), location: None }', cprover_bindings/src/goto_program/expr.rs:1000:9                    
stack backtrace:                                                                                                                                                                                                                     
   0: rust_begin_unwind                                                                                           
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/std/src/panicking.rs:575:5                
   1: core::panicking::panic_fmt                                                                                  
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/panicking.rs:65:14                                                                                                                                  
   2: cprover_bindings::goto_program::expr::Expr::binop                                                                                                                                                                              
             at /kani/cprover_bindings/src/goto_program/expr.rs:1000:9
   3: cprover_bindings::goto_program::expr::Expr::sub_overflow_result                                             
             at /kani/cprover_bindings/src/goto_program/expr.rs:1468:9                                                                                                           
   4: kani_compiler::codegen_cprover_gotoc::codegen::intrinsic::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::codegen_intrinsic                                                                          
             at /kani/kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:631:36                                       
   5: kani_compiler::codegen_cprover_gotoc::codegen::intrinsic::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::codegen_funcall_of_intrinsic                                                               
             at /kani/kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:74:21                                                                                         
   6: kani_compiler::codegen_cprover_gotoc::codegen::statement::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::codegen_funcall                                                                            
             at /kani/kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs:513:20
   7: kani_compiler::codegen_cprover_gotoc::codegen::statement::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::codegen_terminator                                                                         
             at /kani/kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs:209:17
   8: kani_compiler::codegen_cprover_gotoc::codegen::block::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::codegen_block
             at /kani/kani-compiler/src/codegen_cprover_gotoc/codegen/block.rs:22:29                      
   9: kani_compiler::codegen_cprover_gotoc::codegen::function::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::codegen_function::{{closure}}                                                               
             at /kani/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:90:69                                                                                          
  10: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}                                                                                                                                                            
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/iter/traits/iterator.rs:828:29
  11: core::iter::adapters::map::map_fold::{{closure}}                                                                                                                                                                               
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/iter/adapters/map.rs:84:21                                                                                                                          
  12: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold::enumerate::{{closure}}                                                                                                        
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/iter/adapters/enumerate.rs:106:27 
  13: core::iter::traits::iterator::Iterator::fold
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/iter/traits/iterator.rs:2414:21
  14: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/iter/adapters/enumerate.rs:112:9
  15: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/iter/adapters/map.rs:124:9
  16: core::iter::traits::iterator::Iterator::for_each                                                            
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/core/src/iter/traits/iterator.rs:831:9
  17: kani_compiler::codegen_cprover_gotoc::codegen::function::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::codegen_function
             at /kani/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:90:13
  18: <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{{closure}}::{{closure}}
             at /kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:130:39
  19: kani_compiler::codegen_cprover_gotoc::utils::debug::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::call_with_panic_debug_info::{{closure}}
             at /kani/kani-compiler/src/codegen_cprover_gotoc/utils/debug.rs:68:13
  20: std::thread::local::LocalKey<T>::try_with
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/std/src/thread/local.rs:446:16
  21: std::thread::local::LocalKey<T>::with
             at /rustc/c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b/library/std/src/thread/local.rs:422:9
  22: kani_compiler::codegen_cprover_gotoc::utils::debug::<impl kani_compiler::codegen_cprover_gotoc::context::goto_ctx::GotocCtx>::call_with_panic_debug_info
             at /kani/kani-compiler/src/codegen_cprover_gotoc/utils/debug.rs:65:9
  23: <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{{closure}}
             at /kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:129:29
  24: kani_compiler::codegen_cprover_gotoc::compiler_interface::with_timer
             at /kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:518:15
  25: <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
             at /kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:99:9
  26: <rustc_session::session::Session>::time::<alloc::boxed::Box<dyn core::any::Any>, rustc_interface::passes::start_codegen::{closure#0}>
  27: rustc_interface::passes::start_codegen
  28: <rustc_interface::passes::QueryContext>::enter::<<rustc_interface::queries::Queries>::ongoing_codegen::{closure#0}::{closure#0}, core::result::Result<alloc::boxed::Box<dyn core::any::Any>, rustc_errors::ErrorGuaranteed>>
  29: <rustc_interface::queries::Queries>::ongoing_codegen
  30: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
  31: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1
}>::{closure#0}::{closure#1}>


```

### PR DESCRIPTION
**Pull request by JustusAdam**
### Description of changes: 

This PR addresses one of the issue raised in #2121 where a typecheck for `ctpop` in the backend makes kani crash. 

With these changes it now reports a clean compiler error with the usual span information, e.g.

```
error: Type check failed for intrinsic `ctpop`: Expected integer type, found ()
   |
12 |     let n = ctpop(());
   |             ^^^^^^^^^

error: aborting due to previous error
```

### Resolved issues:

Relates to #2121

Concrete coverage of the subissues
- [x] `ctpop`
- [ ] ~`sub_with_overflow`~ (ignored because failure happens in rustc, several similar issues are reported there and it seems there is no consensus whether a compiler panic on intrinsics should be considered appropriate rust-lang/rust#101962, rust-lang/rust#92495)
- [ ] ~`_mm` family~ (not reproducible for me, requires `x86`)

### Call-outs / outstanding questions

- [x] At the moment I essentially perform the same type check that cprover_bindings does but earlier, when we have access to the rustc session and can gracefully emit errors. The clean thing to do here is to only perform this type check in one place. Probably cleanest to have functions that perform type checks (such as Expr::popcont) return Result so that it propagates the error gracefully and then turn it into a rustc error. That's a bigger change though that I think needs feedback.
- [x] I used the Debug printing of the internal Type struct here, since that's what's available at that point and because it does not implement Display, but this won't mean much to the user so I'll have to revisit this to make it more meaningful.
- [x] Emitting errors via tcx.sess does not immediately abort the compilation (unlike assert!). This means I need to emit a no-op Stmt to make codegen_intrinsic's type happy. Depending on the statement used and when exactly the compiler aborts this could (possibly) cause weird error downstream. I'm using Stmt::skip but that may not be the right choice.

### Testing:

TODO

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by JustusAdam** [comment]
Thank you for your comments @celinval. I opened this PR so we have the diff available. 

At the moment I added this logic analogous with the example you posted. I can keep doing that for the other subissues, but my concern is that this both has a chance to miss other cases in which such type checks are necessary and also means that all checks are performed twice (in code gen and binding library) which means they can start to diverge at some point.

Wrt types, I can make sure to use `Ty` instead.

Thanks for clarifying on the abort. I'm not quite sure where the nearest `abort_if_errors` call is to the intrinsics lowering. Are you saying that I should add one?

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 1]
```diff
@@ -445,7 +446,17 @@ impl<'tcx> GotocCtx<'tcx> {
             "cosf64" => codegen_simple_intrinsic!(Cos),
             "ctlz" => codegen_count_intrinsic!(ctlz, true),
             "ctlz_nonzero" => codegen_count_intrinsic!(ctlz, false),
-            "ctpop" => self.codegen_expr_to_place(p, fargs.remove(0).popcount()),
+            "ctpop" => {
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:449]: For readability, can you please move this to a little function?

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 2]
```diff
@@ -445,7 +446,17 @@ impl<'tcx> GotocCtx<'tcx> {
...
+                let arg = fargs.remove(0);
+                let atype = arg.typ();
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:451]: nit: `a_type` or `arg_type`

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 3]
```diff
@@ -445,7 +446,17 @@ impl<'tcx> GotocCtx<'tcx> {
...
+                if !atype.is_integer() {
+                    self.tcx.sess.span_err(span.into_iter().collect::<Vec<_>>(), format!("Type Check Failure: Expected integer type for call to `ctpop`, found {atype:?}"));
+                    Stmt::skip(loc)
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:454]: abort compilation

**Review by celinval — COMMENTED** [review]
Awesome! That's about it. Can you also please add a test?

**Comment by JustusAdam** [comment]
I can do that. What I feel like I haven't quite gotten an answer to yet is what I am supposed to do with the similar issues. #2121 contains a few concrete examples, but presumably this whole thing is bigger, as in there are many intrinsics that have type requirements that are checked via assert in `cprover_bindings` and could cause similar crashes to the ones reported in #2121. Do you want me to just fix the examples from the issue or generally propagate all potential type check errors that may arise from the typer checking in `cprover_bindings`? It feels like this may be the right time to nip these in the bud?

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs (hunk 1 above)
Comment (JustusAdam) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs]: Done

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 4]
```diff
@@ -120,16 +120,71 @@ impl<'tcx> GotocCtx<'tcx> {
         }
     }
 
-    /// c.f. `rustc_codegen_llvm::intrinsic` `impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx>`
-    /// `fn codegen_intrinsic_call`
-    /// c.f. <https://doc.rust-lang.org/std/intrinsics/index.html>
+    fn codegen_ctpop(
+        &mut self,
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:124]: Nit: can you please move these aux functions to be after the codegen_intrinsics?

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 5]
```diff
@@ -120,16 +120,71 @@ impl<'tcx> GotocCtx<'tcx> {
...
+    /// Report that a delayed type check on an intrinsic failed.
+    ///
+    /// The idea is to blame one of the arguments on the failed type check and
+    /// report the type that was found for that argument in `actual`. The
+    /// `expected` type for that argument can be very permissive (e.g. "some
+    /// integer type") and as a result it allows a permissive string as
+    /// description.
+    ///
+    /// Calling this function will abort the compilation though that is not
+    /// obvious by the type.
```
[the remark this answers is not in the document]
Comment (JustusAdam) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:148]: I'm not sure that works. `abort_if_errors` returns `()` not `!` so I think the type checker wouldn't like me much.

File: tests/expected/intrinsics/sub_with_workflow_ice_fixme/arithmetic_zst_fixme.rs
[the remark this answers is not in the document]
Comment (JustusAdam) [reply tests/expected/intrinsics/sub_with_workflow_ice_fixme/arithmetic_zst_fixme.rs:1]: Oops, I meant to only rename the other, not sure what happened here.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 6]
```diff
@@ -120,16 +120,71 @@ impl<'tcx> GotocCtx<'tcx> {
...
     ) -> Stmt {
+        // ### A note on type checking
+        //
+        // The backend/codegen generally assumes that at this point arguments
+        // have been type checked and that the given intrinsic is safe to call
+        // with the provided arguments. However in rare cases the intrinsics
+        // type signature is too permissive or has to be liberal because the
+        // types are enforced by the specific code gen/backend. In such cases we
+        // handle the type checking here. The type constraints are lifted
+        // straight from the assertions made in various functions in
```
[the remark this answers is not in the document]
Comment (JustusAdam) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:183]: I reworded the last part, hopefully it captures that now

**Comment by JustusAdam** [comment]
I force pushed an update with the requested changes to this branch but for some reason I don't see it reflected here. I'll have to investigate that. Sorry for the confusion

**Comment by JustusAdam** [comment]
Okay now it's reflected.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs]: You can add an unreachable after the `abort_if_errors`. :)

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 7]
```diff
@@ -650,6 +661,49 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        actual: Ty,
+    ) {
+        self.tcx.sess.span_err(
+            span.into_iter().collect::<Vec<_>>(),
+            format!(
+                "Type check failed for intrinsic `{name}`: Expected {expected}, found {actual}"
+            ),
+        );
+        self.tcx.sess.abort_if_errors();
+    }
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:705]: I would just add an `unreachable!()` after the abort_if_errors() and return never.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 8]
```diff
@@ -650,6 +661,49 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        &mut self,
+        target_place: Place<'tcx>,
+        loc: Location,
+        span: Option<Span>,
+        arg: Expr,
+        arg_rust_ty: Ty<'tcx>,
+    ) -> Stmt {
+        if !arg.typ().is_integer() {
+            self.intrinsics_typecheck_fail(span, "ctpop", "integer type", arg_rust_ty);
+            Stmt::skip(loc)
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:675]: If you return never, you won't need this skip().

**Review by celinval — APPROVED** [review]

**Review by celinval — APPROVED** [review]

**Comment by adpaco-aws** [comment]
Looks like only a few `rustdoc` errors are blocking this PR... @JustusAdam can you please take a look?

**Comment by celinval** [comment]
BTW, can you please update the title of this PR and description to something more meaningful? Thanks!

**Comment by JustusAdam** [comment]
I'm not sure what the perf regression error is supposed to tell me there. Can anyone shed a light?

**Comment by karkhaz** [comment]
> I'm not sure what the perf regression error is supposed to tell me there. Can anyone shed a light?

That actually looks like it's broken in CI. Don't worry about it, this check isn't blocking the PR.

**Pull request merged**